MREUnityRuntime/MREUnityRuntimeLib/Core/Collider.cs

```diff
@@ -30,56 +30,78 @@ public enum ColliderType
         /// </summary>
         None = 0,
 
+        /// <summary>
+        /// Choose best collider shape for mesh
+        /// </summary>
+        Auto,
+
         /// <summary>
         /// Box shaped collider.
         /// </summary>
-        Box = 1,
+        Box,
 
         /// <summary>
         /// Sphere shaped collider.
         /// </summary>
-        Sphere = 2,
+        Sphere,
 
         /// <summary>
-        /// Mesh based collider.
+        /// Capsule shaped collider.
         /// </summary>
-        Mesh = 3,
+        Capsule,
 
         /// <summary>
-        /// Capsule shaped collider.
+        /// Mesh collider.
         /// </summary>
-        Capsule = 4
+        Mesh
     }
 
     internal class Collider : MonoBehaviour, ICollider
     {
         private UnityCollider _collider;
-        private Actor _ownerActor;
         private ColliderEventType _colliderEventSubscriptions = ColliderEventType.None;
+        private int colliderGeneration = -1;
+
+        private Actor _actor;
+        private Actor Actor
+        {
+            get
+            {
+                if (_actor == null)
+                {
+                    _actor = gameObject.GetComponent<Actor>();
+                }
+                return _actor;
+            }
+        }
 
         /// <inheritdoc />
-        public bool IsEnabled => _collider.enabled;
+        public bool IsEnabled { get; private set; } = true;
 
         /// <inheritdoc />
-        public bool IsTrigger => _collider.isTrigger;
+        public bool IsTrigger { get; private set; } = false;
 
         // /// <inheritdoc />
         //public CollisionLayer CollisionLayer { get; set; }
 
         /// <inheritdoc />
-        public ColliderType ColliderType { get; private set; }
+        public ColliderType Shape { get; private set; }
 
-        internal void Initialize(UnityCollider unityCollider)
-        {
-            _ownerActor = unityCollider.gameObject.GetComponent<Actor>()
-                ?? throw new Exception("An MRE collider must be associated with a Unity game object that is an MRE actor.");
-            _collider = unityCollider;
-        }
+        // cannot be Auto
+        private ColliderType _actualShape;
 
         internal void ApplyPatch(ColliderPatch patch)
         {
-            _collider.enabled = _collider.enabled.GetPatchApplied(IsEnabled.ApplyPatch(patch.IsEnabled));
-            _collider.isTrigger = _collider.isTrigger.GetPatchApplied(IsTrigger.ApplyPatch(patch.IsTrigger));
+            IsEnabled = IsEnabled.GetPatchApplied(IsEnabled.ApplyPatch(patch.IsEnabled));
+            IsTrigger = IsTrigger.GetPatchApplied(IsTrigger.ApplyPatch(patch.IsTrigger));
```

Comment:
I strongly dislike that the values are being maintained in this class separately from the actual collider on the unity game object.  That should always be the authority as it was before the change above.

Reply:
Summary of discussions: keep actor patching in the actor, leave the collider class for collision events only, and make sure that if we have to wait for a mesh, all fields of a patch get applied together.

---
In reply to: [313530051](https://github.com/microsoft/mixed-reality-extension-unity/pull/98#discussion_r313530051) [](ancestors = 313530051)